File: codex_monitor/owner_health.py
"""Bounded, read-only health probes for an existing Codex App Server owner.

The dashboard needs to answer a narrow operational question: can an existing
owner currently receive work for one saved conversation?  This module uses
only explicit WebSocket transports.  It never creates a stdio App Server,
starts or resumes a thread, writes a queue item, or asks a model to run.
"""

from __future__ import annotations

from collections import OrderedDict
import json
import math
import time
from typing import Any, Callable
from urllib.parse import urlsplit


DEFAULT_TIMEOUT = 0.45
DEFAULT_CACHE_TTL = 2.0
DEFAULT_CACHE_SIZE = 64
MAX_PAGES = 8
MAX_MESSAGE_BYTES = 256 * 1024
MAX_THREAD_IDS = 4096

_AUTH_WORDS = (
    "auth", "credential", "forbidden", "invalid token", "logged out",
    "login", "log in", "not authenticated", "unauthorized", "unauthorised",
)
_UNAVAILABLE_WORDS = (
    "unavailable", "connection", "connect", "closed", "disconnect",
    "refused", "timed out", "timeout", "network", "unreachable",
)


class OwnerProbeError(Exception):
    """An internal bounded probe failure; its message is never user output."""

    def __init__(self, kind: str, method: str | None = None):
        self.kind = kind
        self.method = method
        super().__init__(kind)


class _RpcError(Exception):
    def __init__(self, code: Any, message: Any):
        self.code = code
        self.message = str(message or "RPC error")
        super().__init__(self.message)


def _supported_endpoint(endpoint: Any) -> bool:
    if not isinstance(endpoint, str) or not endpoint or any(ord(char) < 32 for char in endpoint):
        return False
    if endpoint.startswith("unix://"):
        path = endpoint.removeprefix("unix://")
        return path.startswith("/") and len(path) > 1 and "\x00" not in path
    if not endpoint.startswith(("ws://", "wss://")):
        return False
    try:
        parsed = urlsplit(endpoint)
        host = parsed.hostname
        parsed.port
    except ValueError:
        return False
    if (
        parsed.scheme not in {"ws", "wss"} or not parsed.netloc or not host or
        parsed.username or parsed.password or parsed.fragment or
        any(char.isspace() for char in endpoint) or parsed.netloc.endswith(":")
    ):
        return False
    return parsed.scheme == "wss" or host.lower() in {"127.0.0.1", "localhost", "::1"}


def _display_endpoint(endpoint: Any) -> str:
    """Keep credentials out of probe observations, including rejected URLs."""

    if not isinstance(endpoint, str):
        return "-"
    try:
        parsed = urlsplit(endpoint)
    except ValueError:
        return "owner endpoint"
    if parsed.username or parsed.password:
        host = parsed.hostname or "owner"
        try:
            parsed_port = parsed.port
        except ValueError:
            parsed_port = None
        port = f":{parsed_port}" if parsed_port is not None else ""
        return f"{parsed.scheme}://{host}{port}{parsed.path}"
    return endpoint[:240]


def _auth_error(error: BaseException) -> bool:
    code = getattr(error, "code", None)
    if code in {401, 403, -32001, -32002}:
        return True
    text = str(error).casefold()
    return any(word in text for word in _AUTH_WORDS) or "http 401" in text or "http 403" in text


def _unsupported(error: BaseException) -> bool:
    if getattr(error, "code", None) == -32601:
        return True
    text = str(error).casefold()
    return "method not found" in text or "unknown variant" in text or "unsupported method" in text


def _status_value(value: Any) -> str | None:
    if isinstance(value, str):
        value = value.strip()
        if value and len(value) <= 64:
            return "".join(char if ord(char) >= 32 and ord(char) != 127 else " " for char in value)
    if isinstance(value, dict):
        for key in ("type", "status", "state"):
            result = _status_value(value.get(key))
            if result:
                return result
    return None


class _OwnerRpc:
    """One short-lived WebSocket RPC connection."""

    def __init__(self, endpoint: str, *, timeout: float, token: str | None):
        try:
            from websockets.sync.client import connect, unix_connect
        except Exception as exc:  # pragma: no cover - dependency is declared
            raise OwnerProbeError("unavailable") from exc
        try:
            if endpoint.startswith("unix://"):
                self._socket = unix_connect(
                    endpoint.removeprefix("unix://"),
                    open_timeout=timeout,
                    close_timeout=min(timeout, 0.5),
                    compression=None,
                    max_size=MAX_MESSAGE_BYTES,
                )
            else:
                headers = {"Authorization": "Bearer " + token} if token else None
                self._socket = connect(
                    endpoint,
                    additional_headers=headers,
                    open_timeout=timeout,
                    close_timeout=min(timeout, 0.5),
                    compression=None,
                    max_size=MAX_MESSAGE_BYTES,
                )
        except Exception as exc:
            raise OwnerProbeError("auth-required" if _auth_error(exc) else "unavailable") from exc
        self._next_id = 0

    def call(self, method: str, params: dict[str, Any], *, timeout: float) -> Any:
        self._next_id += 1
        request_id = self._next_id
        try:
            encoded = json.dumps({"id": request_id, "method": method, "params": params}, separators=(",", ":"))
            if len(encoded.encode("utf-8")) > MAX_MESSAGE_BYTES:
                raise OwnerProbeError("unavailable", method)
            self._socket.send(encoded)
            deadline = time.monotonic() + max(0.01, timeout)
            while True:
                remaining = deadline - time.monotonic()
                if remaining <= 0:
                    raise OwnerProbeError("unavailable", method)
                raw = self._socket.recv(timeout=remaining)
                if raw is None:
                    raise OwnerProbeError("unavailable", method)
                if isinstance(raw, bytes):
                    if len(raw) > MAX_MESSAGE_BYTES:
                        raise OwnerProbeError("unavailable", method)
                    raw = raw.decode("utf-8")
                if not isinstance(raw, str) or len(raw.encode("utf-8")) > MAX_MESSAGE_BYTES:
                    raise OwnerProbeError("unavailable", method)
                try:
                    message = json.loads(raw)
                except (TypeError, ValueError, RecursionError) as exc:
                    raise OwnerProbeError("unavailable", method) from exc
                if not isinstance(message, dict) or message.get("id") != request_id:
                    continue
                if "error" in message and isinstance(message["error"], dict):
                    error = message["error"]
                    raise _RpcError(error.get("code"), error.get("message"))
                if "result" not in message:
                    raise OwnerProbeError("unavailable", method)
                return message["result"]
        except OwnerProbeError:
            raise
        except _RpcError:
            raise
        except Exception as exc:
            raise OwnerProbeError("auth-required" if _auth_error(exc) else "unavailable", method) from exc

    def notify(self, method: str, params: dict[str, Any]) -> None:
        try:
            self._socket.send(json.dumps({"method": method, "params": params}, separators=(",", ":")))
        except Exception as exc:
            raise OwnerProbeError("unavailable", method) from exc

    def close(self) -> None:
        try:
            self._socket.close()
        except Exception:
            pass


def _page(rpc: _OwnerRpc, method: str, *, timeout: float) -> list[Any]:
    rows: list[Any] = []
    cursor: str | None = None
    seen: set[str] = set()
    for _ in range(MAX_PAGES):
        params: dict[str, Any] = {} if cursor is None else {"cursor": cursor}
        result = rpc.call(method, params, timeout=timeout)
        if not isinstance(result, dict) or not isinstance(result.get("data"), list):
            raise OwnerProbeError("unavailable", method)
        rows.extend(result["data"])
        if len(rows) > MAX_THREAD_IDS:
            raise OwnerProbeError("unavailable", method)
        next_cursor = result.get("nextCursor")
        if not next_cursor:
            return rows
        if not isinstance(next_cursor, str) or next_cursor in seen:
            raise OwnerProbeError("unavailable", method)
        seen.add(next_cursor)
        cursor = next_cursor
    raise OwnerProbeError("unavailable", method)


def _account_observation(result: Any) -> dict[str, Any]:
    if not isinstance(result, dict):
        return {
            "status": "unknown",
            "present": None,
            "credential_validation": "unverified",
            "refresh_token": False,
        }
    account = result.get("account") if "account" in result else None
    requires_auth = result.get("requiresOpenaiAuth")
    if account is None and requires_auth is True:
        return {
            "status": "auth-required",
            "present": False,
            "credential_validation": "unverified",
            "refresh_token": False,
        }
    if account is None and requires_auth is not False:
        status = "unknown"
    else:
        status = "present" if account else "absent"
    # Keep this a presence observation.  An account object is not evidence
    # that a credential will authorize a later operation.
    return {
        "status": status,
        "present": bool(account) if status in {"present", "absent"} else None,
        "credential_validation": "unverified",
        "refresh_token": False,
    }


def _base(endpoint: str, thread: str) -> dict[str, Any]:
    return {
        "endpoint": _display_endpoint(endpoint),
        "thread": thread,
        "status": "unverified",
        "state": "unverified",
        "ready": False,
        "transport_reachable": False,
        "thread_loaded": None,
        "account": {"status": "unverified", "present": None, "credential_validation": "unverified"},
        "credential_validation": "unverified",
        "model_execution": "unverified",
        "reason": None,
    }


def probe_owner(
    endpoint: str,
    thread: str,
    *,
    token: str | None = None,
    timeout: float = DEFAULT_TIMEOUT,
    rpc_factory: Callable[..., Any] | None = None,
) -> dict[str, Any]:
    """Probe one existing owner using only bounded, read-only RPC methods."""

    result = _base(endpoint, thread)
    if endpoint in {"shared-local", "local"}:
        result["reason"] = "owner is not explicit; readiness is unverified"
        return result
    if not _supported_endpoint(endpoint):
        result["reason"] = "endpoint is not a supported explicit WebSocket owner"
        return result
    if not isinstance(thread, str) or not thread:
        result["reason"] = "conversation identity is unavailable"
        return result
    if isinstance(timeout, bool) or not isinstance(timeout, (int, float)) or not math.isfinite(timeout):
        timeout = DEFAULT_TIMEOUT
    timeout = max(0.05, min(float(timeout), DEFAULT_TIMEOUT))
    rpc: Any = None
    try:
        rpc = (rpc_factory or _OwnerRpc)(endpoint, timeout=timeout, token=token)
        rpc.call(
            "initialize",
            {"clientInfo": {"name": "codex-monitor-dashboard", "title": "Codex Monitor Dashboard", "version": "0.1.0"},
             "capabilities": {"experimentalApi": True}},
            timeout=timeout,
        )
        rpc.notify("initialized", {})
        result["transport_reachable"] = True

        try:
            account = rpc.call("account/read", {"refreshToken": False}, timeout=timeout)
        except _RpcError as exc:
            if _unsupported(exc):
                result["account"] = {"status": "unsupported", "present": None,
                                      "credential_validation": "unverified", "refresh_token": False}
            elif _auth_error(exc):
                result.update(status="auth-required", state="auth-required", reason="authentication required")
                return result
            else:
                result["account"] = {"status": "unavailable", "present": None,
                                      "credential_validation": "unverified", "refresh_token": False}
        else:
            result["account"] = _account_observation(account)
            if result["account"].get("status") == "auth-required":
                result.update(status="auth-required", state="auth-required", reason="authentication required")
                return result

        try:
            loaded = _page(rpc, "thread/loaded/list", timeout=timeout)
        except _RpcError as exc:
            if _auth_error(exc):
                result.update(status="auth-required", state="auth-required", reason="authentication required")
            else:
                result["reason"] = "loaded conversations could not be read"
            return result
        except OwnerProbeError:
            result["reason"] = "loaded conversations could not be read"
            return result
        loaded_ids = {row if isinstance(row, str) else row.get("id") for row in loaded if isinstance(row, (str, dict))}
        loaded_ids.discard(None)
        result["loaded_threads"] = sorted(loaded_ids)[:MAX_THREAD_IDS]
        result["thread_loaded"] = thread in loaded_ids
        if thread not in loaded_ids:
            result.update(status="unloaded", state="unloaded", reason="conversation is not loaded by this owner")
            return result

        try:
            thread_read = rpc.call(
                "thread/read", {"threadId": thread, "includeTurns": False}, timeout=timeout
            )
        except _RpcError as exc:
            if _auth_error(exc):
                result.update(status="auth-required", state="auth-required", reason="authentication required")
                return result
            if not _unsupported(exc):
                result.update(status="unavailable", state="unavailable", reason="thread status unavailable")
                return result
            # Older peers may lack thread/read, so retain the loaded-list
            # readiness observation with an explicit unsupported note.
            result["thread_read"] = {"status": "unsupported"}
        except OwnerProbeError:
            result.update(status="unavailable", state="unavailable", reason="thread status unavailable")
            return result
        else:
            status = None
            if isinstance(thread_read, dict):
                status = _status_value(thread_read.get("status"))
                if status is None and isinstance(thread_read.get("thread"), dict):
                    status = _status_value(thread_read["thread"].get("status"))
            result["thread_read"] = {"status": status or "observed"}
        result.update(status="ready-to-receive", state="ready-to-receive", ready=True,
                      reason="owner transport reachable and conversation loaded")
        return result
    except _RpcError as exc:
        result.update(status="auth-required" if _auth_error(exc) else "unavailable",
                      state="auth-required" if _auth_error(exc) else "unavailable",
                      reason="authentication required" if _auth_error(exc) else "owner RPC unavailable")
        return result
    except OwnerProbeError as exc:
        result.update(status=exc.kind, state=exc.kind,
                      reason="authentication required" if exc.kind == "auth-required" else "owner unavailable")
        return result
    except Exception:
        result.update(status="unavailable", state="unavailable", reason="owner unavailable")
        return result
    finally:
        if rpc is not None:
            try:
                rpc.close()
            except Exception:
                pass


class OwnerHealthCache:
    """Small TTL cache preventing a live dashboard from reconnecting per frame."""

    def __init__(self, *, probe: Callable[..., dict[str, Any]] = probe_owner,
                 ttl: float = DEFAULT_CACHE_TTL, max_entries: int = DEFAULT_CACHE_SIZE,
                 clock: Callable[[], float] = time.monotonic):
        if not callable(probe):
            raise ValueError("owner probe must be callable")
        if isinstance(ttl, bool) or not isinstance(ttl, (int, float)) or not math.isfinite(ttl) or ttl <= 0:
            raise ValueError("owner health cache ttl must be positive")
        if type(max_entries) is not int or max_entries <= 0:
            raise ValueError("owner health cache size must be positive")
        self.probe = probe
        self.ttl = float(ttl)
        self.max_entries = max_entries
        self.clock = clock
        self._items: OrderedDict[tuple[str, str], tuple[float, dict[str, Any]]] = OrderedDict()

    def get(self, endpoint: str, thread: str) -> dict[str, Any]:
        key = (endpoint, thread)
        now = self.clock()
        cached = self._items.get(key)
        if cached and cached[0] > now:
            self._items.move_to_end(key)
            return dict(cached[1])
        value = self.probe(endpoint, thread)
        if not isinstance(value, dict):
            value = _base(endpoint, thread)
            value.update(status="unavailable", state="unavailable", reason="owner probe returned invalid data")
        self._items[key] = (now + self.ttl, dict(value))
        self._items.move_to_end(key)
        while len(self._items) > self.max_entries:
            self._items.popitem(last=False)
        return dict(value)

File: rust/src/owner_health.rs
//! Bounded, read-only observations of an existing Codex owner.
//!
//! Owner health answers whether an explicitly addressed owner can currently
//! receive work for a loaded conversation. It never starts a local owner,
//! resumes a thread, queues input, or infers model execution.

use codex_monitor_rs::session::{SessionError, SessionPool, validate_endpoint};
use serde_json::{Value, json};
use std::collections::{BTreeSet, HashMap};
use std::sync::{Mutex, OnceLock};
use std::time::Duration;
use tokio::time::{Instant, timeout};

const PROBE_TIMEOUT: Duration = Duration::from_millis(450);
const PROBE_BUDGET: Duration = Duration::from_secs(3);
const CACHE_TTL: Duration = Duration::from_secs(2);
const CACHE_SIZE: usize = 64;
const MAX_PAGES: usize = 8;
const MAX_THREAD_IDS: usize = 4096;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
enum Failure {
    Auth,
    Unsupported,
    Unavailable,
}

#[derive(Clone)]
struct EndpointCache {
    expires_at: Instant,
    base: Value,
    loaded: Option<Vec<String>>,
    threads: HashMap<String, Value>,
}

#[derive(Default)]
struct CacheState {
    entries: HashMap<String, EndpointCache>,
}

pub struct OwnerHealthCache {
    state: Mutex<CacheState>,
}

impl OwnerHealthCache {
    pub fn new() -> Self {
        Self {
            state: Mutex::new(CacheState {
                entries: HashMap::new(),
            }),
        }
    }

    async fn get(
        &self,
        pool: &SessionPool,
        endpoint: &str,
        thread: &str,
        deadline: Instant,
    ) -> Value {
        let cached = self
            .state
            .lock()
            .ok()
            .and_then(|state| state.entries.get(endpoint).cloned())
            .filter(|entry| entry.expires_at > Instant::now());
        let mut entry = if let Some(entry) = cached {
            entry
        } else {
            let mut entry = probe_endpoint(pool, endpoint, deadline).await;
            entry.expires_at = Instant::now() + CACHE_TTL;
            self.insert(endpoint, entry.clone());
            entry
        };

        if let Some(value) = entry.threads.get(thread) {
            return value.clone();
        }
        let Some(loaded) = entry.loaded.as_ref() else {
            return with_thread(&entry.base, thread);
        };
        if !loaded.iter().any(|value| value == thread) {
            let value = unloaded(&entry.base, thread, loaded);
            entry.threads.insert(thread.to_owned(), value.clone());
            self.insert(endpoint, entry);
            return value;
        }

        let value = probe_thread(pool, endpoint, thread, &entry.base, loaded, deadline).await;
        entry.threads.insert(thread.to_owned(), value.clone());
        self.insert(endpoint, entry);
        value
    }

    fn insert(&self, endpoint: &str, entry: EndpointCache) {
        let Ok(mut state) = self.state.lock() else {
            return;
        };
        state.entries.insert(endpoint.to_owned(), entry);
        while state.entries.len() > CACHE_SIZE {
            let Some(oldest) = state
                .entries
                .iter()
                .min_by_key(|(_, value)| value.expires_at)
                .map(|(key, _)| key.clone())
            else {
                break;
            };
            state.entries.remove(&oldest);
        }
    }
}

static GLOBAL_CACHE: OnceLock<OwnerHealthCache> = OnceLock::new();

pub fn global_cache() -> &'static OwnerHealthCache {
    GLOBAL_CACHE.get_or_init(OwnerHealthCache::new)
}

pub async fn attach(bindings: &mut Value, pool: &SessionPool) {
    let deadline = Instant::now() + PROBE_BUDGET;
    let Some(rows) = bindings.as_array_mut() else {
        return;
    };
    for binding in rows {
        let thread = binding.get("thread").and_then(Value::as_str).unwrap_or("");
        let endpoint = binding
            .get("endpoint")
            .and_then(Value::as_str)
            .unwrap_or("");
        let reason = if binding["removed"] == true {
            Some("route is removed")
        } else if binding["enabled"] != true {
            Some("route is disabled")
        } else if thread.is_empty() {
            Some("conversation identity is unavailable")
        } else if !supported_endpoint(endpoint) {
            Some("owner endpoint is not an explicit supported transport")
        } else if Instant::now() >= deadline {
            Some("owner probe budget exhausted")
        } else {
            None
        };
        binding["owner_health"] = if let Some(reason) = reason {
            base(endpoint, thread, reason)
        } else {
            global_cache().get(pool, endpoint, thread, deadline).await
        };
    }
}

fn supported_endpoint(endpoint: &str) -> bool {
    if endpoint == "shared-local" || endpoint == "local" || endpoint.starts_with("ssh://") {
        return false;
    }
    if endpoint == "unix:///" || endpoint.starts_with("unix:///") && endpoint.len() <= 9 {
        return false;
    }
    (endpoint.starts_with("ws://")
        || endpoint.starts_with("wss://")
        || endpoint.starts_with("unix://"))
        && validate_endpoint(endpoint).is_ok()
}

fn display_endpoint(endpoint: &str) -> String {
    let parsed = url::Url::parse(endpoint);
    let Ok(parsed) = parsed else {
        return endpoint.chars().take(240).collect();
    };
    if parsed.username().is_empty() && parsed.password().is_none() {
        return endpoint.chars().take(240).collect();
    }
    let host = parsed.host_str().unwrap_or("owner");
    let port = parsed
        .port()
        .map(|port| format!(":{port}"))
        .unwrap_or_default();
    format!("{}://{}{}{}", parsed.scheme(), host, port, parsed.path())
}

fn base(endpoint: &str, thread: &str, reason: &str) -> Value {
    json!({
        "endpoint": display_endpoint(endpoint),
        "thread": thread,
        "status": "unverified",
        "state": "unverified",
        "ready": false,
        "transport_reachable": false,
        "thread_loaded": Value::Null,
        "account": {"status":"unverified","present":Value::Null,"credential_validation":"unverified","refresh_token":false},
        "credential_validation": "unverified",
        "model_execution": "unverified",
        "reason": reason,
    })
}

fn with_thread(base: &Value, thread: &str) -> Value {
    let mut value = base.clone();
    value["thread"] = json!(thread);
    value
}

fn terminal(base: &Value, thread: &str, status: &str, reason: &str) -> Value {
    let mut value = with_thread(base, thread);
    value["status"] = json!(status);
    value["state"] = json!(status);
    value["ready"] = json!(false);
    value["reason"] = json!(reason);
    value
}

async fn probe_endpoint(pool: &SessionPool, endpoint: &str, deadline: Instant) -> EndpointCache {
    let mut base = base(endpoint, "", "owner unavailable");
    let account = match call(
        pool,
        endpoint,
        "account/read",
        json!({"refreshToken":false}),
        deadline,
    )
    .await
    {
        Ok(value) => {
            base["transport_reachable"] = json!(true);
            account_observation(&value)
        }
        Err(Failure::Auth) => {
            base["transport_reachable"] = json!(true);
            base["account"] = account_auth_required();
            return EndpointCache {
                expires_at: Instant::now(),
                base: terminal(&base, "", "auth-required", "authentication required"),
                loaded: None,
                threads: HashMap::new(),
            };
        }
        Err(Failure::Unsupported) => {
            base["transport_reachable"] = json!(true);
            json!({"status":"unsupported","present":Value::Null,"credential_validation":"unverified","refresh_token":false})
        }
        Err(Failure::Unavailable) => {
            return EndpointCache {
                expires_at: Instant::now(),
                base: terminal(&base, "", "unavailable", "owner unavailable"),
                loaded: None,
                threads: HashMap::new(),
            };
        }
    };
    base["account"] = account.clone();
    if account["status"] == "auth-required" {
        return EndpointCache {
            expires_at: Instant::now(),
            base: terminal(&base, "", "auth-required", "authentication required"),
            loaded: None,
            threads: HashMap::new(),
        };
    }

    let loaded = match loaded_pages(pool, endpoint, deadline).await {
        Ok(value) => value,
        Err(Failure::Auth) => {
            return EndpointCache {
                expires_at: Instant::now(),
                base: terminal(&base, "", "auth-required", "authentication required"),
                loaded: None,
                threads: HashMap::new(),
            };
        }
        Err(_) => {
            base["reason"] = json!("loaded conversations could not be read");
            return EndpointCache {
                expires_at: Instant::now(),
                base,
                loaded: None,
                threads: HashMap::new(),
            };
        }
    };
    base["transport_reachable"] = json!(true);
    EndpointCache {
        expires_at: Instant::now(),
        base,
        loaded: Some(loaded),
        threads: HashMap::new(),
    }
}

async fn probe_thread(
    pool: &SessionPool,
    endpoint: &str,
    thread: &str,
    base: &Value,
    loaded: &[String],
    deadline: Instant,
) -> Value {
    match call(
        pool,
        endpoint,
        "thread/read",
        json!({"threadId":thread,"includeTurns":false}),
        deadline,
    )
    .await
    {
        Ok(value) => {
            let mut result = with_thread(base, thread);
            result["loaded_threads"] = json!(loaded);
            result["thread_loaded"] = json!(true);
            result["thread_read"] =
                json!({"status": status_value(&value).unwrap_or_else(|| "observed".into())});
            result["status"] = json!("ready-to-receive");
            result["state"] = json!("ready-to-receive");
            result["ready"] = json!(true);
            result["reason"] = json!("owner transport reachable and conversation loaded");
            result
        }
        Err(Failure::Unsupported) => {
            let mut result = with_thread(base, thread);
            result["loaded_threads"] = json!(loaded);
            result["thread_loaded"] = json!(true);
            result["thread_read"] = json!({"status":"unsupported"});
            result["status"] = json!("ready-to-receive");
            result["state"] = json!("ready-to-receive");
            result["ready"] = json!(true);
            result["reason"] = json!("owner transport reachable and conversation loaded");
            result
        }
        Err(Failure::Auth) => thread_failure(
            base,
            thread,
            loaded,
            "auth-required",
            "authentication required",
        ),
        Err(Failure::Unavailable) => thread_failure(
            base,
            thread,
            loaded,
            "unavailable",
            "thread status unavailable",
        ),
    }
}

async fn loaded_pages(
    pool: &SessionPool,
    endpoint: &str,
    deadline: Instant,
) -> Result<Vec<String>, Failure> {
    let mut rows = BTreeSet::new();
    let mut cursor: Option<String> = None;
    let mut seen = BTreeSet::new();
    for _ in 0..MAX_PAGES {
        let params = cursor
            .as_ref()
            .map_or_else(|| json!({}), |value| json!({"cursor":value}));
        let value = call(pool, endpoint, "thread/loaded/list", params, deadline).await?;
        let Some(data) = value.get("data").and_then(Value::as_array) else {
            return Err(Failure::Unavailable);
        };
        for row in data {
            let id = row
                .as_str()
                .or_else(|| row.get("id").and_then(Value::as_str));
            if let Some(id) = id {
                rows.insert(id.to_owned());
            }
        }
        if rows.len() > MAX_THREAD_IDS {
            return Err(Failure::Unavailable);
        }
        let next = value.get("nextCursor");
        if next.is_none() || next == Some(&Value::Null) {
            return Ok(rows.into_iter().collect());
        }
        let Some(next) = next.and_then(Value::as_str) else {
            return Err(Failure::Unavailable);
        };
        if !seen.insert(next.to_owned()) {
            return Err(Failure::Unavailable);
        }
        cursor = Some(next.to_owned());
    }
    Err(Failure::Unavailable)
}

async fn call(
    pool: &SessionPool,
    endpoint: &str,
    method: &str,
    params: Value,
    deadline: Instant,
) -> Result<Value, Failure> {
    let remaining = deadline
        .saturating_duration_since(Instant::now())
        .min(PROBE_TIMEOUT);
    if remaining.is_zero() {
        return Err(Failure::Unavailable);
    }
    match timeout(remaining, pool.call(endpoint, method, params)).await {
        Ok(Ok(value)) => Ok(value),
        Ok(Err(error)) => Err(classify_error(&error)),
        Err(_) => Err(Failure::Unavailable),
    }
}

fn classify_error(error: &SessionError) -> Failure {
    let text = error.to_string().to_lowercase();
    if [
        "auth",
        "credential",
        "forbidden",
        "invalid token",
        "logged out",
        "login",
        "unauthorized",
    ]
    .iter()
    .any(|word| text.contains(word))
    {
        Failure::Auth
    } else if text.contains("method not found")
        || text.contains("unsupported")
        || text.contains("unknown variant")
        || text.contains("rejected: ") && text.contains("method")
    {
        Failure::Unsupported
    } else {
        Failure::Unavailable
    }
}

fn account_auth_required() -> Value {
    json!({"status":"auth-required","present":false,"credential_validation":"unverified","refresh_token":false})
}

fn account_observation(value: &Value) -> Value {
    let Some(object) = value.as_object() else {
        return json!({"status":"unknown","present":Value::Null,"credential_validation":"unverified","refresh_token":false});
    };
    let account = object.get("account");
    let requires_auth = object.get("requiresOpenaiAuth").and_then(Value::as_bool);
    if account.is_none_or(Value::is_null) && requires_auth == Some(true) {
        return account_auth_required();
    }
    if account.is_none_or(Value::is_null) && requires_auth != Some(false) {
        return json!({"status":"unknown","present":Value::Null,"credential_validation":"unverified","refresh_token":false});
    }
    let present = account.is_some_and(truthy);
    json!({"status":if present {"present"} else {"absent"},"present":present,"credential_validation":"unverified","refresh_token":false})
}

fn truthy(value: &Value) -> bool {
    match value {
        Value::Null => false,
        Value::Bool(value) => *value,
        Value::Number(value) => value.as_f64().is_some_and(|number| number != 0.0),
        Value::String(value) => !value.is_empty(),
        Value::Array(value) => !value.is_empty(),
        Value::Object(value) => !value.is_empty(),
    }
}

fn status_value(value: &Value) -> Option<String> {
    if let Some(value) = value.as_str() {
        let value = value.trim();
        if !value.is_empty() && value.len() <= 64 {
            return Some(
                value
                    .chars()
                    .map(|character| {
                        if character.is_control() {
                            ' '
                        } else {
                            character
                        }
                    })
                    .collect(),
            );
        }
    }
    value.as_object().and_then(|object| {
        ["type", "status", "state"]
            .iter()
            .find_map(|key| object.get(*key).and_then(status_value))
    })
}

fn unloaded(base: &Value, thread: &str, loaded: &[String]) -> Value {
    let mut value = terminal(
        base,
        thread,
        "unloaded",
        "conversation is not loaded by this owner",
    );
    value["loaded_threads"] = json!(loaded);
    value["thread_loaded"] = json!(false);
    value
}

fn thread_failure(
    base: &Value,
    thread: &str,
    loaded: &[String],
    status: &str,
    reason: &str,
) -> Value {
    let mut value = terminal(base, thread, status, reason);
    value["loaded_threads"] = json!(loaded);
    value["thread_loaded"] = json!(true);
    value
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn only_explicit_supported_owner_transports_are_probeable() {
        assert!(supported_endpoint("ws://127.0.0.1:8765"));
        assert!(supported_endpoint("wss://owner.example"));
        assert!(supported_endpoint("unix:///tmp/owner.sock"));
        assert!(!supported_endpoint("shared-local"));
        assert!(!supported_endpoint("local"));
        assert!(!supported_endpoint("ssh://owner"));
        assert!(!supported_endpoint("ws://example.com:8765"));
        assert!(!supported_endpoint("wss://user:secret@owner.example"));
    }

    #[test]
    fn account_presence_never_claims_credential_validation() {
        let value = account_observation(&json!({"account":{"type":"chatgpt"}}));
        assert_eq!(value["status"], "present");
        assert_eq!(value["credential_validation"], "unverified");
        assert_eq!(value["refresh_token"], false);
        let auth = account_observation(&json!({"account":null,"requiresOpenaiAuth":true}));
        assert_eq!(auth["status"], "auth-required");
        assert_eq!(auth["present"], false);
        let unknown = account_observation(&json!({}));
        assert_eq!(unknown["status"], "unknown");
        assert!(unknown["present"].is_null());
    }

    #[test]
    fn endpoint_display_redacts_credentials() {
        let value = base(
            "wss://user:secret@owner.example:8765/path",
            "thread",
            "test",
        );
        assert_eq!(value["endpoint"], "wss://owner.example:8765/path");
        assert!(!value.to_string().contains("secret"));
    }
}
